In [29]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [30]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending_data = pd.read_csv(
    Path('Resources/lending_data.csv')   
)

# Review the DataFrame
df_lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [31]:
# Separate the data into labels and features

# Separate the y variable, the labels
y_var = df_lending_data['loan_status']

# Separate the X variable, the features
x_var = df_lending_data.drop(columns=['loan_status'])

In [32]:
# Review the y variable Series
y_var.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [33]:
# Review the X variable DataFrame
x_var.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [34]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(x_var, y_var, random_state=1, stratify=y_var)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [35]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [36]:
# Make a prediction using the testing data
testing_data_prediction = lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [46]:
# Generate a confusion matrix for the model
matrix_model = confusion_matrix(y_test, testing_data_prediction)
matrix_df = pd.DataFrame(matrix_model, 
                         index=["Actual 0", "Actual 1"], 
                         columns=["Predicted 0", "Predicted 1"])

print("Confusion Matrix:\n", matrix_df)

Confusion Matrix:
           Predicted 0  Predicted 1
Actual 0        18679           80
Actual 1           67          558


In [47]:
# Print the classification report for the model
class_report = classification_report(y_test, testing_data_prediction)
print("Classification Report:\n", class_report)

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
The logistic regression model demonstrates outstanding performance in predicting healthy loans (0), achieving perfect precision and nearly perfect recall. However, its performance in predicting high-risk loans (1) is comparatively lower, with slightly reduced precision and recall.

The overall and individual F1-scores for this model are all considered to be acceptable harmonic means (1.00, 0.88, 0.94, 0.99). This indicates the model's effective balance of recall and precision in its predictions. Additionally, the model shows no difficulty in detecting True Negatives (100%), though it identifies True Positives with a slightly lower precision (85%). Overall, this model exhibits very high accuracy (99%) with exceptional performance in accurately predicting healthy loans. It also maintains a good balance between precision and recall for high-risk loans, though there is some room for improvement in precision specifically for high-risk predictions.

---